### ISSUE TITLE
python.lang.maintainability.useless-inner-function false positive on nested function

### ISSUE THREAD

**Issue by emcd** [returntocorp/sgrep-rules#2367]
**Describe the bug**

Sometimes, it is useful to define decorator factories in Python to customize decorators. This can result in nested functions, such as the greatly simplified example given in the `To Reproduce` section of this report.

Running Semgrep on these nested functions results in the following false positive:
```
$ semgrep --config .local/scm-modules/semgrep-rules/python/lang --error --use-git-ignore semgrep_bug.py
Scanning 1 file with 158 python rules.

Findings:

  semgrep_bug.py
     local.scm-modules.semgrep-rules.python.lang.maintainability.useless-inner-function
        function `function_wrapper` is defined inside a function but never used

          3┆ def function_wrapper( *args, **kwargs ):
          4┆     # Do something with 'foo'.
          5┆     return function( *args, **kwargs )
```

**To Reproduce**

1. Create a file with the following content:
```python
def decorator_factory( foo ):
    def decorator( function ):
        def function_wrapper( *args, **kwargs ):
            # Do something with 'foo'.
            return function( *args, **kwargs )
        return function_wrapper
    return decorator

@decorator_factory( 'bar' )
def test( ): ''' Simple reproducer. '''
```
2. Run Semgrep with the `python.lang.maintainability.useless-inner-function` on it.

**Expected behavior**

No error. This is a legitimate pattern and the inner function is not. in fact, useless.

**Priority**
How important is this to you?
- [ ] P0: blocking me from making progress
- [ ] P1: this will block me in the near future
- [X] P2: annoying but not blocking me

I can use `# nosemgrep` for now but would prefer to have it scan the function.

**Additional context**
Could be related to #660, but most of the links in that report are either inaccessible to me or broken, so I don't know for certain. Also, this report has the correct spelling of `useless-inner-function`, which should be easier for other people to find.

**Comment by p4p3r** [comment]
Hello @emcd and thanks for reporting this issue!

This looks to be indeed a duplicate of #660.
I added your reproducer code as a new FP test case (https://github.com/returntocorp/semgrep-rules/pull/2368) and resumed the conversation on the existing issue comment thread.
I'm going to close this issue as a duplicate for now.

Let me know if you have any question!
